res/controllers/Traktor-Kontrol-S4-MK2-hid-scripts.js

```diff
@@ -0,0 +1,1277 @@
+  engine.setValue(group, "pregain", cur_pregain + delta);
+}
+
+TraktorS4MK2.callbackLoopMove = function(field) {
+  // TODO: common-hid-packet-parser looks like it should do deltas, but I can't get them to work.
+  var splitted = field.id.split(".");
+  var group = splitted[0]
+  prev_loopmove = TraktorS4MK2.controller.prev_loopmove[group];
+  TraktorS4MK2.controller.prev_loopmove[group] = field.value;
+  var delta = 0;
+  if (prev_loopmove === 15 && field.value === 0) {
+    delta = 1;
+  } else if (prev_loopmove === 0 && field.value === 15) {
+    delta = -1;
+  } else if (field.value > prev_loopmove) {
+    delta = 1;
+  } else {
+    delta = -1;
+  }
+
+  engine.setValue(field.group, "loop_move", delta);
+}
+
+TraktorS4MK2.callbackLoopSize = function(field) {
+  var splitted = field.id.split(".");
+  var group = splitted[0]
+  prev_loopsize = TraktorS4MK2.controller.prev_loopsize[group];
+  TraktorS4MK2.controller.prev_loopsize[group] = field.value;
+  var delta = 0;
+  if (prev_loopsize === 15 && field.value === 0) {
+    engine.setValue(field.group, "loop_double", 1);
+    engine.setValue(field.group, "loop_double", 0);
+  } else if (prev_loopsize === 0 && field.value === 15) {
+    engine.setValue(field.group, "loop_halve", 1);
+    engine.setValue(field.group, "loop_halve", 0);
+  } else if (field.value > prev_loopsize) {
+    engine.setValue(field.group, "loop_double", 1);
+    engine.setValue(field.group, "loop_double", 0);
+  } else {
+    engine.setValue(field.group, "loop_halve", 1);
+    engine.setValue(field.group, "loop_halve", 0);
+  }
+}
+
+TraktorS4MK2.callbackBrowse = function(field) {
+  // TODO: common-hid-packet-parser looks like it should do deltas, but I can't get them to work.
+  prev_browse = TraktorS4MK2.controller.prev_browse;
+  TraktorS4MK2.controller.prev_browse = field.value;
+  var delta = 0;
+  if (prev_browse === 15 && field.value === 0) {
+    delta = 1;
+  } else if (prev_browse === 0 && field.value === 15) {
+    delta = -1;
+  } else if (field.value > prev_browse) {
+    delta = 1;
+  } else {
+    delta = -1;
+  }
+
+  engine.setValue("[Playlist]", "SelectTrackKnob", delta);
+}
+
+TraktorS4MK2.scalerParameter = function(group, name, value) {
+  return script.absoluteLin(value, 0, 1, 16, 4080);
+}
+
+TraktorS4MK2.scalerVolume = function(group, name, value) {
+  if (group === "[Master]") {
+    return script.absoluteNonLin(value, 0, 1, 4, 16, 4080);
+  } else {
+    return script.absoluteNonLin(value, 0, 0.25, 1, 16, 4080);
+  }
+}
+
+TraktorS4MK2.scalerQuickKnob = function(group, name, value) {
+  return script.absoluteLin(value, 0, 1, 16, 4080);
+}
+
+TraktorS4MK2.scalerSlider = function(group, name, value) {
+  return script.absoluteLin(value, -1, 1, 16, 4080);
+}
+
+TraktorS4MK2.resolveDeckIfActive = function(group) {
+  var controller = TraktorS4MK2.controller;
+  if (group === "[Channel1]") {
+    if (controller.left_deck_C) {
+      return undefined;
```

Comment:
in JS land, the difference between null and undefined is vague -- but undefined conventionally means a variable has not been given a value whereas null means the variable has the null value. So it's a little strange to return undefined. Because undefined is also a value (provided to allow you to test if something is undefined) you can use it like this :). 

Just mentioning it -- I think what you have works fine.